crates/red_knot/src/types.rs

```diff
@@ -46,6 +51,45 @@ impl Type {
     pub const fn is_unknown(&self) -> bool {
         matches!(self, Type::Unknown)
     }
+
+    // NOTE: naming? `get_member_type`
+    pub fn get_member(&self, db: &dyn SemanticDb, name: &Name) -> QueryResult<Type> {
+        // NOTE: maybe make this a trait that each TypeID implements?
+        match self {
+            Type::Any => todo!("attribute lookup on Any type"),
+            Type::Never => todo!("attribute lookup on Never type"),
+            Type::Unknown => todo!("attribute lookup on Unknown type"),
+            Type::Unbound => todo!("attribute lookup on Unbound type"),
+            Type::Function(_) => todo!("attribute lookup on Function type"),
+            Type::Module(module_id) => module_id
+                .get_member(db, name)
+                // FIXME: don't want to unwrap here; need to add a QueryError ctor for the case
+                // where the get_member result is Ok(None)
+                .map(Option::unwrap),
+            Type::Class(class_id) => class_id
+                .get_own_class_member(db, name)
+                .or_else(|_| class_id.get_super_class_member(db, name))
+                // FIXME: don't want to unwrap here; need to add a QueryError ctor for the case
+                // where the get_member result is Ok(None)
+                .map(Option::unwrap),
+            Type::Instance(_) => {
+                // TODO MRO? get_own_instance_member, get_instance_member
+                todo!("attribute lookup on Instance type")
+            }
+            Type::Union(union_id) => {
+                let jar: &SemanticJar = db.jar()?;
+                let union_ref = jar.type_store.get_union(*union_id);
+                // TODO return a type FF at least one of the unioned-types have the member; return
+                // the union over those member types
```

Comment:
Accessing a member of a union should really be a type error if any of the unioned types lack the member. In inference-only mode we won't want to issue this error, though -- I think we should add `Unknown` to the member type union instead.

Reply:
Long-term it would be interesting to still return a "guessed" type when only some variants implement the specific member. It could allow for better recovery in some situations (e.g. intelli sense). The API would need to be explicit that the type isn't certain.

Reply:
Yes, that's actually what my comment proposes, that e.g. if we have `x: A | B | C` and we access `x.foo`, and types `A` and `B` have a `foo` attribute but `C` does not, we would return `A.foo | B.foo | Unknown` instead of erroring. Effectively this is our best guess of the type, and the union with `Unknown` is the marker of our uncertainty.

Reply:
Yeah this is a good question, and I don't have a clear answer yet. I need to think through more carefully when we should return None vs Unknown, and how that will interact with inference-only vs type-checking modes.

In principle a nice clean answer would to always use Unknown and never return None. But I worry that this doesn't allow us to distinguish "attribute doesn't exist" from "attribute exists but resolving its type gives us Unknown." But maybe it's ok to not distinguish these from the caller, if we're able to surface diagnostics wherever we create an Unknown. Which I think we should be able to. (It may require passing some kind of diagnostic-sink to all these methods.)

Reply:
On the other hand I can imagine that in some lint rules we might want to be able to distinguish these cases, without necessarily creating a diagnostic for the missing attribute.

Reply:
I think that keeping those cases distinct is important at this level, but perhaps it should wait until we have a use case to drive their clarification.